### PR TITLE
upstream snmp sonic-yang files

### LINKED ISSUE #10549: [yang] Need Yang for SNMP table
#### Description

We don't have Yang model for SNMP table.

#### Steps to reproduce the issue:
1. Run 'config apply-patch' command
2. We can get below info:
sonic_yang(6):Note: Below table(s) have no YANG models: CONSOLE_SWITCH, DEVICE_NEIGHBOR_METADATA, DHCP_SERVER, KDUMP, RESTAPI, SNMP, SNMP_COMMUNITY, TELEMETRY

#### Describe the results you received:
sonic_yang reports no Yang models.

#### Describe the results you expected:
sonic_yang reports ok.

#### Output of `show version`:

```
(paste your output here)
```
#### Output of `show techsupport`:

```
(paste your output here or download and attach the file here )
```

### LINKED ISSUE #10550: [yang] Need Yang for SNMP_COMMUNITY table
#### Description

We don't have Yang model for SNMP_COMMUNITY table.

#### Steps to reproduce the issue:
1. Run 'config apply-patch' command
2. We can get below info:
sonic_yang(6):Note: Below table(s) have no YANG models: CONSOLE_SWITCH, DEVICE_NEIGHBOR_METADATA, DHCP_SERVER, KDUMP, RESTAPI, SNMP, SNMP_COMMUNITY, TELEMETRY

#### Describe the results you received:
sonic_yang reports no Yang models.

#### Describe the results you expected:
sonic_yang reports ok.

#### Output of `show version`:

```
(paste your output here)
```
#### Output of `show techsupport`:

```
(paste your output here or download and attach the file here )
```

### PR DESCRIPTION
**Pull request by suresh-rupanagudi**
Fix #10549
Fix #10550

#### Why I did it
Create sonic yang model for SNMP
Tables:SNMP, SNMP_COMMUNITY
#### How I did it
Defined yang models based for SNMP based on snmp.yml
#### How to verify it
Added test cases to verify

### CODE REVIEW

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/snmp.json [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                }
+            }
+        }
+    },
+
+    "SNMP_COMMUNITY_TEST": {
+        "sonic-snmp:sonic-snmp": {
+            "sonic-snmp:SNMP_COMMUNITY": {
+                "SNMP_COMMUNITY_LIST": [
+                {
```
Comment (ganglyu) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/snmp.json:37]: Please check indent

File: src/sonic-yang-models/yang-models/sonic-snmp.yang [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    container SNMP {
+      container CONTACT {
+        leaf Contact {
+          type string {
+             length "1..255";
+          }
+          description
+            "SNMP System Contact.";
+        }
+	  }
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-snmp.yang:33]: Please replace tab

File: src/sonic-yang-models/yang-models/sonic-snmp.yang [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+
+  container sonic-snmp {
+
+    container SNMP {
+      container CONTACT {
+        leaf Contact {
+          type string {
+             length "1..255";
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-snmp.yang:28]: why not 0?

**Review by ganglyu — APPROVED** [review]

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/snmp.json (hunk 1 above)
Comment (suresh-rupanagudi) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/snmp.json]: taken care. Please see new changes

File: src/sonic-yang-models/yang-models/sonic-snmp.yang (hunk 2 above)
Comment (suresh-rupanagudi) [reply src/sonic-yang-models/yang-models/sonic-snmp.yang]: Taken care.

File: src/sonic-yang-models/tests/files/sample_config_db.json [hunk 4]
```diff
@@ -1800,7 +1800,24 @@
...
              }
-         }
+         },
+
+        "SNMP": {
+            "CONTACT": {
+                "Contact": "sonic-support@xxx.com"
+            },
+            "LOCATION": {
+                "Location": "Sonic Location"
```
Comment (qiluo-msft) [inline src/sonic-yang-models/tests/files/sample_config_db.json:1810]: >Sonic [](http://example.com/codeflow?start=29&length=5)

It is not invented by SONiC. Let's call it "SNMP Server Location"

File: src/sonic-yang-models/tests/files/sample_config_db.json [hunk 5]
```diff
@@ -1800,7 +1800,24 @@
                 "num_dumps": "3",
                 "memory": "0M-2G:256M,2G-4G:256M,4G-8G:384M,8G-:448M"
              }
-         }
+         },
+
+        "SNMP": {
+            "CONTACT": {
+                "Contact": "sonic-support@xxx.com"
```
Comment (qiluo-msft) [inline src/sonic-yang-models/tests/files/sample_config_db.json:1807]: >sonic-support@xxx.com [](http://example.com/codeflow?start=28&length=21)

Let's use `testuser@contoso.com` which is more safe.

**Review by ganglyu — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Comment by qiluo-msft** [comment]
Could you help resolve the conflict, and merge latest master if not following it?

**Comment by qiluo-msft** [comment]
Could you merge master again and resolve the conflicts?

**Comment by wen587** [comment]
Hi @suresh-rupanagudi , I saw the test always fail on GCU test.
```
Error: Given patch will produce invalid config. Error: Data Loading Failed\nAll Keys are not parsed in SNMP_COMMUNITY\ndict_keys(['public'])\nexceptionList:[\"'TYPE'\"]",
```
Here is the reason:
In your defined YANG MODEL, the leaf `type` is lowercase. https://github.com/Azure/sonic-buildimage/blob/8bd08d2658bf28524cec08da7138ba630c22f4dd/src/sonic-yang-models/yang-models/sonic-snmp.yang#L63

But in SONiC default config, the `type` is by default uppercase. Below is the example of my SONiC config.
```
    "SNMP_COMMUNITY": {
        "public": {
            "TYPE": "RO"                    <=uppercase `type`
        }
    }
```

I think the default SONiC SNMP config need to be updated.

**Comment by suresh-rupanagudi** [comment]
Thanks JingWeng, Gang for looking into this.
1)why are we not getting this error when we build "sonic-broadcom.bin" locally?.
What needs to be done to run this locally?.

Now solution is 
   1)to change "type" to "TYPE" in yang file. OR
   2)change "TYPE" to "type" in snmp_yml_to_configdb.py file. We might also need to take care all the tests where community is used.  
Yang standard allows defining identifier with capital letters.
But I could not see any yang module using this. 
Please let me know 

From rfc8407.
===========
"Identifiers SHOULD follow a consistent naming pattern throughout the
   module.  Only lowercase letters, numbers, and dashes SHOULD be used
   in identifier names.  Uppercase characters, the period character, and
   the underscore character MAY be used if the identifier represents a
   well-known value that uses these characters.  YANG does not permit
   any other characters in YANG identifiers."

**Comment by ganglyu** [comment]
Pipeline will run basic test, so we can detect errors before merging PR.
Please follow sonic-mgmt/docs at master * sonic-net/sonic-mgmt (github.com)<https://github.com/sonic-net/sonic-mgmt/tree/master/docs> to setup testbed and run tests.

I think modify yang file will be easier, and sonic-acl.yang is using capital letters.

BR
Gang

**Comment by wen587** [comment]
> Thanks JingWeng, Gang for looking into this. 1)why are we not getting this error when we build "sonic-broadcom.bin" locally?. What needs to be done to run this locally?.

1)Because it doesn't have any test failures during the image build. 
It failed during the sonic-mgmt test, because the actual config is not as that as we defined.

For example, you define the SNMP_COMMUNITY['public']['type'] with RO or RW, so you just test the config [here](https://github.com/sonic-net/sonic-buildimage/pull/10828/files#diff-35a7f6a55896179bc30a6aa51e5eaf7bc06cf3f541b6c77cfeef0e270f59fd7fR1910) with 
```
        "SNMP_COMMUNITY": {
            "public": {
                "type": "RO"           <===== lowercase 'type'
            },
``` 

However it tested real machine during sonic-mgmt test and found that the real config doesn't validate SNMP_COMMUNITY's YANG. Because the real config is 
```
    "SNMP_COMMUNITY": {
        "public": {
            "TYPE": "RO"             <===== uppercase 'TYPE'
        }
    },
``` 
> What needs to be done to run this locally?.

I think you can modify [the config here](https://github.com/sonic-net/sonic-buildimage/pull/10828/files#diff-35a7f6a55896179bc30a6aa51e5eaf7bc06cf3f541b6c77cfeef0e270f59fd7fR1912-R1915) and change `type` to `TYPE`. It also need related YANG model update.

**Review by wen587 — APPROVED** [review]

**Comment by suresh-rupanagudi** [comment]
Hi jingwenxie,
Can you please look at failures.

**Comment by yxieca** [comment]
@suresh-rupanagudi this change cannot be cherrypicked to 202205 cleanly. Please create new PR for 202205 branch.

**Comment by yxieca** [comment]
@suresh-rupanagudi, @zhangyanzhao, can you help create separate PR for 202205 branch? If the change has been done. Please link the 202205 PR to this PR so that I can close the cherry-picking request.

**Comment by suresh-rupanagudi** [comment]
ETA for PR on 202205  branch is 2/20/2023

**Comment by yxieca** [comment]
@suresh-rupanagudi is this PR not needed for 202211 branch?

File: src/sonic-yang-models/yang-models/sonic-snmp.yang [hunk 6]
```diff
@@ -0,0 +1,74 @@
+module sonic-snmp {
+  namespace "http://github.com/Azure/sonic-snmp";
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-snmp.yang:2]: Please use sonic-net to replace Azure.

**Comment by suresh-rupanagudi** [comment]
Hi,
Created new PR on 202205
https://github.com/sonic-net/sonic-buildimage/pull/13896

Not sure if this PR required for 202211 branch. Please cherry pick changes if required.

**Pull request merged**